overlord/ifacestate/ifacestate.go

```diff
@@ -35,17 +36,90 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
-var noConflictOnConnectTasks = func(task *state.Task) bool {
-	return task.Kind() != "connect" && task.Kind() != "disconnect"
+func noConflictOnConnectTasks(chg *state.Change) func(task *state.Task) bool {
+	checkConflict := func(task *state.Task) bool {
+		if chg != nil && task.Change() != nil {
+			// if same change, then return false (no conflict)
+			return chg.ID() != task.Change().ID()
+		}
+		return task.Kind() != "connect" && task.Kind() != "disconnect"
+	}
+	return checkConflict
+}
+
+var connectRetryTimeout = time.Second * 5
+
+func checkConnectConflicts(st *state.State, change *state.Change, plugSnap, slotSnap string) error {
+	for _, chg := range st.Changes() {
+		if chg.Status().Ready() || (change != nil && change.ID() == chg.ID()) {
+			continue
+		}
+		if chg.Kind() == "transition-ubuntu-core" {
+			return fmt.Errorf("ubuntu-core to core transition in progress, no other changes allowed until this is done")
+		}
+	}
+
+	for _, task := range st.Tasks() {
+		k := task.Kind()
+		chg := task.Change()
+
+		// there is no conflict if same change
+		if change != nil && chg != nil && change.ID() == chg.ID() {
+			continue
+		}
+
+		if k == "connect" || k == "disconnect" {
+			continue
+		}
+
+		// task already executed, no conflict
+		if task.Status() == state.DoneStatus || task.Status() == state.ErrorStatus {
+			continue
+		}
+
+		snapsup, err := snapstate.TaskSnapSetup(task)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot obtain snap setup from task: %s", task.Summary())
+		}
+
+		snapName := snapsup.Name()
+		// different snaps - no conflict
+		if snapName != plugSnap && snapName != slotSnap {
+			continue
+		}
+
+		switch k {
+		case "unlink-snap":
+			// snap getting removed - conflict
+			var snap string
+			if snapName == plugSnap {
+				snap = plugSnap
+			} else {
+				snap = slotSnap
+			}
+			return snapstate.ChangeConflictError(snap, task.Change().Kind())
```

Comment:
I think for auto-connect we want to Retry as well here, the effect is just that we will not even find the snap after some cycles of waiting, is just not relevant, not an error, for auto-connect